libs/nss.c

```diff
@@ -91,57 +91,80 @@ static FILE *tcb_safe_open(const char *file, const char *name)
 	return f;
 }
 
-int _nss_tcb_getspnam_r(const char *name, struct spwd *__result_buf,
-    char *__buffer, size_t __buflen, struct spwd **__result)
+enum nss_status _nss_tcb_getspnam_r(const char *name,
+    struct spwd *__result_buf, char *__buffer, size_t __buflen, int *__errnop)
 {
 	FILE *f;
 	char *file;
 	int retval, saved_errno;
 
 	/* Disallow potentially-malicious user names */
 	if (!is_valid_username(name)) {
-		errno = ENOENT;
+		/* we don't serve an entry here */
+		*__errnop = ENOENT;
 		return NSS_STATUS_NOTFOUND;
 	}
 
-	if (asprintf(&file, TCB_FMT, name) < 0)
+	if (asprintf(&file, TCB_FMT, name) < 0) {
+		/* retry, as malloc or another resource has failed */
+		*__errnop = EAGAIN;
 		return NSS_STATUS_TRYAGAIN;
+	}
+
 	f = tcb_safe_open(file, name);
 	free(file);
-	if (!f)
-		return errno == ENOENT ? NSS_STATUS_NOTFOUND : NSS_STATUS_UNAVAIL;
+	if (!f) {
+		/* $user/shadow not existing nor readable */
+		*__errnop = ENOENT;
+		return NSS_STATUS_UNAVAIL;
+	}
 
-	retval = fgetspent_r(f, __result_buf, __buffer, __buflen, __result);
+	retval = fgetspent_r(f, __result_buf, __buffer,
+	    __buflen, &__result_buf);
 	saved_errno = errno;
 	fclose(f);
 	errno = saved_errno;
-	if (!retval)
-		return NSS_STATUS_SUCCESS;
 
-	switch (saved_errno) {
+	/* real error number is retval from fgetspent_r(),
+	   by NSS spec errnop *MUST NOT* be set to 0 */
+	if (retval)
+		*__errnop = retval;
+
+	switch (retval) {
```

Comment:
This is correct, the error code returned by `fgetspent_r` is currently either `ENOENT` or `EAGAIN`, it doesn't necessarily have to match the value of `errno`.